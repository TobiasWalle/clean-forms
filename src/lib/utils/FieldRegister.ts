import { createDebouncedFunction } from './createDebouncedFunction';
import { removeItemFromArray } from './removeItemFromArray';

export type Path = string[];
export type Paths = Path[];

export interface FieldRegisterChanges {
  registered: Paths;
  unregistered: Paths;
}

export type FieldRegisterListener = (changes: FieldRegisterChanges) => void;

export class FieldRegister {
  private listeners: FieldRegisterListener[] = [];
  private changes: FieldRegisterChanges;

  private _paths: Paths = [];
  public get paths(): Paths {
    return this._paths;
  }

  private triggerListeners = createDebouncedFunction(() => {
    this.listeners.forEach(listener => listener(this.changes));
    this.resetChanges();
  }, 10);

  constructor() {
    this.resetChanges();
  }

  public includesPath(path: Path): boolean {
    return this.paths.includes(path);
  }

  public register(path: Path): void {
    this._paths.push(path);
    this.changes.registered.push(path);
    this.triggerListeners();
  }

  public unregister(path: Path): void {
    removeItemFromArray(this.paths, path);
    this.changes.unregistered.push(path);
    this.triggerListeners();
  }

  public addListener(callback: FieldRegisterListener): void {
    this.triggerListenerIfPathsAlreadyRegistered(callback);
    this.listeners.push(callback);
  }

  private triggerListenerIfPathsAlreadyRegistered(callback: FieldRegisterListener): void {
    if (this.paths.length > 0) {
      callback({ registered: this.paths, unregistered: [] });
    }
  }

  public removeListener(callback: FieldRegisterListener): void {
    removeItemFromArray(this.listeners, callback);
  }

  private resetChanges() {
    this.changes = { registered: [], unregistered: [] };
  }
}
