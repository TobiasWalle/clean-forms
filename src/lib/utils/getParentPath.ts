import { Path } from './FieldRegister';

export function getParentPath(path: Path): Path {
  const length = path.length;
  if (length <= 1) {
    throw new ReferenceError('Path does not have a parent');
  }
  return path.slice(0, length - 1);
}
