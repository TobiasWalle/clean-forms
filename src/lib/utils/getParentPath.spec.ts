import { Path } from './FieldRegister';
import { getParentPath } from './getParentPath';

describe('getParentPath', () => {
  it('should get parent', () => {
    const path = ['a', 'b', 'c'];
    const expectedParent = ['a', 'b'];

    const result = getParentPath(path);

    expect(result).toEqual(expectedParent);
  });

  it('should throw error if path has a size of 1', () => {
    const path = ['a'];

    expect(() => getParentPath(path)).toThrowErrorMatchingSnapshot();
  });

  it('should throw error if path is empty', () => {
    const path: Path = [];

    expect(() => getParentPath(path)).toThrowErrorMatchingSnapshot();
  });
});
