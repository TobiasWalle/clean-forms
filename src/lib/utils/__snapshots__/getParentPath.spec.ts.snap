// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getParentPath should throw error if path has a size of 1 1`] = `"Path does not have a parent"`;

exports[`getParentPath should throw error if path is empty 1`] = `"Path does not have a parent"`;
